packages/belos/tpetra/test/BlockCG/test_bl_cg_hb_multiprec.cpp

```diff
@@ -41,120 +41,126 @@
-int mptestdim, numrhs;
-double *dvals;
+int dim, numrhs; 
+float *fvals;
 int *colptr, *rowind;
 int mptestmypid = 0;
-int mptestnumimages = 1;
 
 ///////////////////////////////////////////////////////////////////////////
-///////////////////////////////////////////////////////////////////////////
-template <class Scalar>
-RCP<LinearProblem<Scalar,Tpetra::MultiVector<Scalar>,Tpetra::Operator<Scalar> > > buildProblem(const RCP<Map<> >& vmap)
+template <class Scalar> 
+RCP<LinearProblem<Scalar,MultiVector<Scalar,int>,Operator<Scalar,int> > > buildProblem()
 {
-  typedef ScalarTraits<Scalar>            SCT;
-  typedef Tpetra::Operator<Scalar>        OP;
-  typedef Tpetra::MultiVector<Scalar>     MV;
-  typedef OperatorTraits<Scalar,MV,OP>    OPT;
-  typedef MultiVecTraits<Scalar,MV>       MVT;
-  using GO = typename MV::global_ordinal_type;
-
-  bool precond = true;
-  RCP<CrsMatrix<Scalar> > A = rcp(new CrsMatrix<Scalar>(vmap,rnnzmax));
-  RCP<CrsMatrix<Scalar> > P = rcp(new CrsMatrix<Scalar>(vmap,1));  // Diagonal preconditioner
+  using SCT = typename ScalarTraits<Scalar>;
+  using MT = typename SCT::magnitudeType;
+  using OP = typename Operator<Scalar,int>;
+  using MV = typename MultiVector<Scalar,int>;
+  using OPT = typename OperatorTraits<Scalar,MV,OP>;
+  using MVT = typename MultiVecTraits<Scalar,MV>;
+
+  RCP<CrsMatrix<Scalar,int> > A = rcp(new CrsMatrix<Scalar,int>(*vmap,rnnzmax));
   if (mptestmypid == 0) {
     // HB format is compressed column. CrsMatrix is compressed row.
-    const double *dptr = dvals;
+    const float *fptr = fvals;
     const int *rptr = rowind;
-    for (int c=0; c<mptestdim; ++c) {
+    for (int c=0; c<dim; ++c) {
       for (int colnnz=0; colnnz < colptr[c+1]-colptr[c]; ++colnnz) {
-        A->insertGlobalValues(*rptr-1,tuple<GO>(c),tuple<Scalar>(*dptr));
-        if (c != *rptr -1)
-          A->insertGlobalValues(c,tuple<GO>(*rptr-1),tuple<Scalar>(*dptr));
-        else
-        {
-          // If the diagonal values are above the safe minimum, then we can use the diagonal preconditioner
-          if ( (*dptr) > SCT::sfmin())
-            P->insertGlobalValues(c,tuple<GO>(*rptr-1),tuple<Scalar>(SCT::one()/(*dptr)));
-          else
-            precond = false;
-        }
-
+        A->insertGlobalValues(*rptr-1,tuple(c),tuple<Scalar>(*fptr));
+        A->insertGlobalValues(c,tuple(*rptr-1),tuple<Scalar>(*fptr));
         ++rptr;
-        ++dptr;
+        ++fptr;
       }
     }
   }
-  // distribute matrix data to other nodes
+  // Distribute matrix data to other nodes
   A->fillComplete();
-  P->fillComplete();
   // Create initial MV and solution MV
   RCP<MV> B, X;
-  X = rcp( new MV(vmap,numrhs) );
-  MVT::MvRandom( *X );
-  B = rcp( new MV(vmap,numrhs) );
+  X = rcp( new MV(*vmap,numrhs) );
+  // Set LHS to actX
+  {
+    typename MultiVector<Scalar,int>::double_pointer Xvals = X->extractView2D();
+    typename MultiVector<float,int>::const_double_pointer actXvals = actX->extractConstView2D();
```

Comment:
Hum, I would have guessed that the correct type here is `Scalar` since that is the type of Xvals?

Reply:
Yes, fixed.